tools/fuzzer/expr_gen.h

```diff
@@ -122,8 +125,19 @@ struct GenConfig {
       {1.0f, 0.1f},  // TypeKind::PointerType
       {1.0f, 0.1f},  // TypeKind::VoidPointerType
   }};
+};
+
+/*
+ * The purpose for generating the type in question. The purpose of this enum
+ * is purely to assist in unit testing.
+ */
+enum class GenTypePurpose {
```

Comment:
It's quite sad that something required just for testing changes the "real" code so much.

It seems, at the moment it's only needed for deducing the result type of the ternary operator. How about we just add the `result_type_` field to the `TernaryExpr`? Then in tests you can just propagate it like this and avoid all this implicit/explicit type picking:

```
// clang-format off
Expr expected = UnaryExpr(
    UnOp::BitNot,
    ParenthesizedExpr(TernaryExpr(
        ScalarType::SignedInt,
        IntegerConstant(1),
        IntegerConstant(2),
        IntegerConstant(3))));
    // clang-format on
```

This would also make sense for other constructs, where you need to deduce result type.